src/storage/src/hummock/event_handler/uploader.rs

```diff
@@ -660,14 +660,23 @@ impl HummockUploader {
             .push_front(imm);
     }
 
-    pub(crate) fn seal_epoch(&mut self, epoch: HummockEpoch) {
-        info_in_release!("epoch {} is sealed", epoch);
-        assert!(
-            epoch > self.max_sealed_epoch,
-            "sealing a sealed epoch {}. {}",
-            epoch,
-            self.max_sealed_epoch
-        );
+    /// Seal an epoch, returns `true` if there's nothing to seal.
+    pub(crate) fn seal_epoch(&mut self, epoch: HummockEpoch) -> bool {
+        match self.max_sealed_epoch.cmp(&epoch) {
+            std::cmp::Ordering::Less => info_in_release!(epoch, "epoch is sealed"),
+            std::cmp::Ordering::Equal => {
+                // This is possible when the cluster recovers.
+                tracing::warn!(epoch, "sealing a sealed epoch, are we recovering?");
+                return true;
+            }
```

Comment:
On recovery, we'll call `seal_epoch` for the `prev_epoch` of the first barrier, which is supposed to have no data and behaves similarly to the very first `INVALID_EPOCH`. So we have to skip it carefully to avoid breaking any assertion.

Since I'm not that familiar with this module, would appreciate your advice! cc @hzxa21 @wenym1

Reply:
Oops, I just reverted the changes in Hummock and now decide whether to skip sealing & syncing in the logic of `LocalBarrierManager`. By doing this ,we can get rid of the special path of `INVALID_EPOCH`. Let's see whether this works. 🤔